dali/python/nvidia/dali/shared_mem.py

```diff
@@ -0,0 +1,108 @@
+You can transfer access to the same shared memory chunk by sending related
+file descriptor available at fd property. Use SharedMem.allocate
+to allocate new chunk of shared memory and SharedMem.open if
+you received file descriptor to already existing memory chunk.
+
+Since Python3.8 there's out of the box support for shared memory, though
+the only way there to transfer the memory to other processes is via filename,
+which might 'leak' if process is closed abruptly.
+
+Parameters
+----------
+`fd` : int
+    File descriptor identifying related shared memory object. Pass -1 to allocate new memory chunk.
+`size` : int
+    When fd=-1 it is the size of shared memory to allocate in bytes, otherwise it must be
+    the size of shared memory objects that provided fd represents.
+"""
+
+    def __init__(self, fd, size):
+        self.shm = _b.SharedMem(fd, size)
+
+    def __getattr__(self, key):
+        # lazily evaluate and cache 'buf' property, so that it is created only once and only when requested
+        if key == 'buf':
+            buf = self.shm.buf()
+            self.__dict__['buf'] = buf
+            return buf
+        raise AttributeError
+
+    @classmethod
+    def allocate(cls, size):
+        """Creates new SharedMem instance representing freshly allocated
+        shared memory of ``size`` bytes.
+
+        Parameters
+        ----------
+        `size` : int
+            Number of bytes to allocate.
+        """
+        return cls(-1, size)
+
+    @classmethod
+    def open(cls, fd, size):
+        """Creates new SharedMem instance that points to already allocated shared
+        memory chunk accessible via provided file descriptor ``fd``.
+
+        Parameters
+        ----------
+        `fd`: int
+            File descriptor pointing to already existing shared memory chunk.
+        `size` : int
+            Size of the existing shared memory chunk.
+        """
+        return cls(fd, size)
+
+    @property
+    def fd(self):
+        """File descriptor, use it to transfer access to the shared memory object to another process.
+        You can transfer it between processes via socket using multiprocessing.reduction.sendfds
+        """
+        return self.shm.fd
+
+    def resize(self, size, trunc=False):
+        """Resize already allocated shared memory chunk. If you want to resize the underlying
+        shared memory chunk pass trunc=True, if the memory chunk has already been resized
+        via another SharedMem instance (possibly in another process), pass new size and
+        trunc=False to simply adjust mmaping of the memory into the current process address space.
+        """
+        if 'buf' in self.__dict__:
+            del self.__dict__['buf']
+        self.shm.resize(size, trunc)
+
+    def close_fd(self):
+        """Close file descriptor identifying memory chunk. You can access the memory
+        via buf property after closing fd, but you neither can transfer the access via fd
```

Comment:
What happens when all processes close their fd? Will it be still accessible?

Reply:
It should be accessible till everybody close fds and unmaps. 

man says ```After a call to mmap(2) the file descriptor may be closed without affecting the memory mapping.```

Reply:
ok